src/main/io/osd.c

```diff
@@ -469,12 +474,33 @@ static bool osdDrawSingleElement(uint8_t item)
 #endif // GPS
 
     case OSD_COMPASS_BAR:
-        memcpy(buff, compassBar + osdGetHeadingIntoDiscreteDirections(DECIDEGREES_TO_DEGREES(attitude.values.yaw), 16), 9);
+      {
+        int16_t h = DECIDEGREES_TO_DEGREES(attitude.values.yaw);
+
+        #ifdef USE_GPS
+          // If MAG_Heading is not available, yaw starts with zero in a random direction. With GPS course over ground the flight direction is roughly displayed instead
+          if (STATE(GPS_FIX) && STATE(GPS_FIX_HOME) && !sensors(SENSOR_MAG | SENSOR_RANGEFINDER)) {
+              h = DECIDEGREES_TO_DEGREES(gpsSol.groundCourse);
+          }
+        #endif
+
+        memcpy(buff, compassBar + osdGetHeadingIntoDiscreteDirections(h, 16), 9);
         buff[9] = 0;
+
         break;
+      }
 
     case OSD_ALTITUDE:
-        osdFormatAltitudeString(buff, getEstimatedAltitude(), true);
+        #ifdef USE_GPS
+          // If BARO is not available, the estimated altitude is zero. Using GPS altitude an absolute height can be displayed instead.
+          if (STATE(GPS_FIX) && !sensors(SENSOR_BARO | SENSOR_RANGEFINDER)) {
```

Comment:
Again, this should not be added in a makeshift way into OSD only, leaving all of the other places to use differing information in a misleading way, this should go into `getEstimatedAltitude()`.

Reply:
Ooh! I'am afraid s0up confusingly started with that old PR. I advised him to restart locally file based against my latest ZIP. That should be clean.
Sorry for that inconvenience. I hope he can avoid the pitfalls I stepped into.